src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -5263,14 +5263,13 @@ private static Class<?>[] insertArgumentsChecks(MethodHandle target, int insCoun
      *                  or if the new method handle's type would have too many parameters
      */
     public static MethodHandle dropArguments(MethodHandle target, int pos, List<Class<?>> valueTypes) {
-        return dropArguments0(target, pos, copyTypes(valueTypes.toArray()));
+        return dropArguments(target, pos, valueTypes.toArray(new Class<?>[0]), false);
```

Comment:
A bit unfortunate that we can't trust this `toArray` to do a copy. I was going to suggest Stream, but it has the same issue (someone might have a custom stream implementation).

I do think a manual copy of the array is possible by having a loop though:
```suggestion
        Class<?>[] ptypes = new Class<?>[valueTypes.size()];
        for (int i = 0; i < ptypes.length; i++) {
            ptypes[i] = valueTypes.get(i);
        }
        return dropArguments(target, pos, ptypes, false);
```

(or, maybe extract such a loop to a helper method for clarity).

Reply:
The same could be done for the public `dropArgumentsToMatch` I think.

Reply:
I'm not so sure. 

First of all we're no worse than before with the defensive copying here. Second of an optimizing compiler might theoretically be able to see that the array we get from the toArray is always fresh and not escaping anywhere in any well-behaved collection, so the clone could be elided. But if not then both toArray and clone are intrinsified operations that are heavily optimized and pretty fast even when interpreting, so emulating it with an external loop might end up taking more time even at peak. While likely taking longer to reach that peak. Using dumb, shared and common code (especially things that get JITted early anyhow) is nice in areas that see most action during startup/warmup.

Reply:
Ok, please keep it the way it is currently then.

Reply:
Correction: `toArray` isn't intrinsified, but common `List` implementations are likely to implement it with a `System.arraycopy`, which is.